### PR TITLE
Agent: Changes required for running webview in clients

### PR DESCRIPTION
**Pull request by pkukielka**
Context: https://sourcegraph.slack.com/archives/C06R69BC8UW/p1723199277787079

## Changes

A bunch of changes needed for getting chat webview to work in other clients, e.g JetBrains, Visual Studio and Eclipse.

Related PR: https://github.com/sourcegraph/jetbrains/pull/2022

## Test plan

Full QA on jetbrains branch

### CODE REVIEW

File: agent/src/agent.ts [hunk 1]
```diff
@@ -1130,17 +1151,11 @@ export class Agent extends MessageHandler implements ExtensionClient {
...
-            const panelId = await this.createChatPanel(
-                Promise.resolve({
-                    type: 'chat',
-                    session: await vscode.commands.executeCommand('cody.chat.newEditorPanel'),
-                })
-            )
-
-            const chatId = this.webPanels.panels.get(panelId)?.chatID ?? ''
-            return { panelId, chatId }
+            await vscode.commands.executeCommand('cody.chat.newEditorPanel')
```
Comment (pkukielka) [inline agent/src/agent.ts:1154]: Won't that break any other client?

**Review by abeatrix — APPROVED** [review]
Left some comments in line and the tests need fixing, but the rest look good to me.

I tested the build for VS Code and everything looks the same as we currently do

image

I noticed there is a bug in the chat webview where clicking on the New Chat button in the sidebar would start a new Editor Chat (where I would expect it to clear my current sidebar chat since there is a move to editor button), but this is reproducible on main, unrelated to this PR. I've reported these bug on Slack here: https://sourcegraph.slack.com/archives/C05MW2TMYAV/p1723207420306119

File: vscode/src/jsonrpc/agent-protocol.ts [hunk 2]
```diff
@@ -525,6 +594,17 @@ export interface ClientCapabilities {
...
+    // AgentWebViewPanel which just delegates bidirectional postMessage over
+    // the Agent protocol. If 'native', implements a larger subset of the VSCode
+    // WebView API and expects the client to run web content in the webview.
+    // Defaults to 'agentic'.
+    webview?: 'agentic' | 'native'
+    // If webview === 'native', describes how the client has configured webview resources.
+    // cspSource is passed to the extension as the Webview cspSource property.
+    // webviewBundleServingPrefix is prepended to resource paths under 'dist' in
+    // asWebviewUri (note, multiple prefixes are not yet implemented.)
+    webviewNativeConfig?: { cspSource: string; webviewBundleServingPrefix: string }
```
Comment (abeatrix) [inline vscode/src/jsonrpc/agent-protocol.ts:607]:
```suggestion
    webview?: 'agentic' | 'native' | undefined | null
    // If webview === 'native', describes how the client has configured webview resources.
    // cspSource is passed to the extension as the Webview cspSource property.
    // webviewBundleServingPrefix is prepended to resource paths under 'dist' in
    // asWebviewUri (note, multiple prefixes are not yet implemented.)
    webviewNativeConfig?: { cspSource: string; webviewBundleServingPrefix: string } | undefined | null
```

I think these are causing the tests to fail?

File: agent/package.json [hunk 3]
```diff
@@ -13,7 +13,8 @@
   "scripts": {
     "build:root": "pnpm -C .. run -s build",
     "build:agent": "node src/esbuild.mjs",
-    "build": "pnpm run -s build:root && pnpm run -s build:agent",
+    "build:webviews": "pnpm -C ../vscode _build:webviews --mode production --outDir $PWD/dist/webviews",
```
Comment (abeatrix) [inline agent/package.json:16]:
```suggestion
    "build:webviews": "pnpm -C ../vscode _build:webviews --mode production --outDir ../../agent/dist/webviews",
```

Running this created a folder in my vscode directory called "$PWD/dist/webviews" 🤣

File: agent/src/agent.ts [hunk 4]
```diff
@@ -1130,17 +1151,11 @@ export class Agent extends MessageHandler implements ExtensionClient {
...
+            return { panelId: 'TODO-remove-panel-id', chatId: 'TODO-remove-chat-id' }
```
[the remark this answers is not in the document]
Comment (abeatrix) [reply agent/src/agent.ts]: i've reverted this change since the panelId here is just TODO.

File: agent/src/agent.ts (hunk 1 above)
Comment (abeatrix) [reply agent/src/agent.ts]: i've reverted this change since the panelId here is just TODO.

File: .npmrc [hunk 5]
```diff
@@ -1 +1,11 @@
 hoist=false
+1
```
Comment (olafurpg) [inline .npmrc:2]: Was this intentional? @pkukielka

Comment (olafurpg) [reply .npmrc:2]: Reverted in https://github.com/sourcegraph/cody/pull/5172

**Pull request merged**